libbeat/common/fifo/fifo.go

```diff
@@ -0,0 +1,84 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package fifo
```

Comment:
This could have some basic unit tests.

Also, potentially could be replaced by https://pkg.go.dev/container/list

Reply:
The reason for not using list (and the reason we have so many similar implementations throughout the repo) is that the stdlib list makes everything an `interface{}`, so there is both extra heap allocations / memory indirection for every node and a type assertion for every buffer access.

Unit tests sound good though, working on that :-)

Reply:
Gah I though they updated it to use generics, I temporarily forgot `any` is just `interface{}`.